msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -93,39 +224,168 @@ public void onTaskCompleted(List<ICacheRecord> cacheRecords) {
+            }
+
+            @Override
+            public void onCancel() {
+                authenticationCallback.onCancel();
+            }
+        };
+    }
+
+    private boolean didExistingCurrentAccountChange(MultiTenantAccount newAccount){
+        MultiTenantAccount persistedAccount = getPersistedCurrentAccount();
+        return (persistedAccount != null || newAccount.getId() != persistedAccount.getId()) ? true : false;
+
+    }
+
+    @Override
+    public void signOut(@NonNull final TaskCompletedCallbackWithError<Boolean, Exception> callback){
+        final String methodName = ":signOut";
         final PublicClientApplicationConfiguration configuration = getConfiguration();
 
-        new BrokerMsalController().removeAccountFromSharedDevice(
-                configuration,
-                new TaskCompletedCallbackWithError<Boolean, Exception>() {
-                    @Override
-                    public void onTaskCompleted(Boolean success) {
-                        if (success) {
-                            persistCurrentAccount(null);
-                        }
 
-                        callback.onTaskCompleted(success);
-                    }
+        try{
+            if (MSALControllerFactory.brokerEligible(
+                    configuration.getAppContext(),
+                    configuration.getDefaultAuthority(),
+                    configuration) && mIsSharedDevice) {
+                removeAccountFromSharedDevice(callback, configuration);
+                return;
+            }
+
+
+            ApiDispatcher.initializeDiagnosticContext();
+
+            MultiTenantAccount persistedCurrentAccount = getPersistedCurrentAccount();
+
+            if (persistedCurrentAccount != null) {
+                final OperationParameters params = OperationParametersAdapter.createOperationParameters(mPublicClientConfiguration);
+                final AccountRecord requestAccountRecord = new AccountRecord();
+                requestAccountRecord.setEnvironment(persistedCurrentAccount.getEnvironment());
+                requestAccountRecord.setHomeAccountId(persistedCurrentAccount.getHomeAccountId());
+                params.setAccount(requestAccountRecord);
+
+                final RemoveAccountCommand command = new RemoveAccountCommand(
+                        params,
+                        MSALControllerFactory.getAcquireTokenController(
+                                mPublicClientConfiguration.getAppContext(),
+                                params.getAuthority(),
+                                mPublicClientConfiguration
+                        ),
+                        callback
+                );
+
+                ApiDispatcher.removeAccount(command);
+            } else {
+                callback.onError(new MsalClientException(MsalClientException.NO_ACCOUNT_TO_SIGN_OUT, "No account is currently signed in to your Single Account Public Client Application"));
```

Comment:
What would be the upgrade path for apps using ADAL who wants to switch to use single account mode?